common/cuda_hip/matrix/dense_kernels.hpp.inc

```diff
@@ -33,6 +33,167 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 namespace kernel {
 
 
+template <typename ValueType, typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void count_nonzero_blocks_per_row(
+        size_type num_block_rows, size_type num_block_cols, size_type stride,
+        int block_size, const ValueType* __restrict__ source,
+        IndexType* __restrict__ block_row_nnz)
+{
+    const auto brow =
+        thread::get_subwarp_id_flat<config::warp_size, IndexType>();
+
+    if (brow >= num_block_rows) {
+        return;
+    }
+
+    const auto bs_sq = block_size * block_size;
+    const auto num_cols = num_block_cols * block_size;
+    auto warp =
+        group::tiled_partition<config::warp_size>(group::this_thread_block());
+    const auto lane = static_cast<IndexType>(warp.thread_rank());
+    constexpr auto full_mask = ~config::lane_mask_type{};
+    constexpr auto one_mask = config::lane_mask_type{1};
+    bool first_block_nonzero = false;
+    IndexType block_count{};
+    for (IndexType base_col = 0; base_col < num_cols;
+         base_col += config::warp_size) {
+        const auto col = base_col + lane;
+        const auto block_local_col = col % block_size;
+        // which is the first column in the current block?
+        const auto block_base_col = col - block_local_col;
+        // collect nonzero bitmask
+        bool local_nonzero = false;
+        for (int local_row = 0; local_row < block_size; local_row++) {
+            const auto row = local_row + brow * block_size;
+            local_nonzero |=
+                col < num_cols && is_nonzero(source[row * stride + col]);
+        }
+        auto nonzero_mask =
+            warp.ballot(local_nonzero) | (first_block_nonzero ? 1u : 0u);
+        // only consider threads in the current block
+        const auto first_thread = block_base_col - base_col;
+        const auto last_thread = first_thread + block_size;
+        // HIP compiles these assertions in Release, traps unconditionally
+        // assert(first_thread < int(config::warp_size));
+        // assert(last_thread >= 0);
+        // mask off everything below first_thread
+        const auto lower_mask =
+            first_thread < 0 ? full_mask : ~((one_mask << first_thread) - 1u);
+        // mask off everything from last_thread
+        const auto upper_mask = last_thread >= config::warp_size
+                                    ? full_mask
+                                    : ((one_mask << last_thread) - 1u);
+        const auto block_mask = upper_mask & lower_mask;
+        const auto local_mask = nonzero_mask & block_mask;
+        // last column in the block increments the counter
+        block_count +=
+            (block_local_col == block_size - 1 && local_mask) ? 1 : 0;
```

Comment:
```suggestion
        block_count +=
            static_cast<IndexType>(block_local_col == block_size - 1 && local_mask);
```
if want to reduce the branch

Reply:
ternary operations usually don't get translated into branches, and this one specifically doesn't even need a cmov, because it is just a written-out conversion bool -> int for clarity.